### ISSUE TITLE
Functionality Roadmap

### ISSUE THREAD

**Issue by seanmor5** [elixir-nx/axon#1]
An issue to track some baseline functionality:

## Activations

- [x] `celu`
- [x] `elu`
- [x] `exp`
- [x] `gelu`
- [x] `hard_tanh`
- [x] `hard_sigmoid`
- [x] `hard_silu`/`hard_swish`
- [x] `leaky_relu`
- [x] `log_sigmoid`
- [x] `relu`
- [x] `relu6`
- [x] `selu`
- [x] `sigmoid`
- [x] `silu`
- [x] `softmax`
- [x] `softplus`
- [x] `softsign`
- [x] `tanh`

## Initializers

- [x] `glorot_uniform`
- [x] `glorot_normal`
- [x] `he_normal`
- [x] `he_uniform`
- [x] `lecun_uniform`
- [x] `lecun_normal`
- [x] `normal`
- [x] `ones`
- [x] `orthogonal` - requires https://github.com/elixir-nx/nx/issues/174
- [x] `uniform`
- [x] `zeros`

## Loss Functions

- [x] `binary_crossentropy`
- [x] `categorical_crossentropy`
- [x] `categorical_hinge`
- [ ] `cosine_similarity` - requires https://github.com/elixir-nx/nx/issues/174
- [x] `ctc`
- [x] `hinge`
- [x] `kl_divergence`
- [x] `log_cosh`
- [x] `margin_ranking`
- [x] `mean_absolute_error`
- [x] `mean_squared_error`
- [x] `poisson`
- [x] `soft_margin`

## Metrics

- [x] `accuracy`
- [x] `mean_squared_error` - requires `defndelegate`
- [x] `mean_absolute_error` - requires `defndelegate`
- [x] `precision`
- [x] `recall`
- [x] `sensitivity`
- [x] `specificty`

## Optimizers

Optax style transformations:

- [x] `scale`
- [x] `scale_by_adam`
- [x] `scale_by_rss`
- [x] `scale_by_belief`
- [x] `scale_by_rms`
- [x] `trace`
- [x] `clip`
- [x] `clip_by_global_norm`
- [x] `centralize`
- [x] `scale_by_trust_ratio`
- [x] `scale_by_schedule`
- [x] `scale_by_radam`
- [x] `scale_by_stddev`

## Schedules

- [x] `polynomial_schedule`
- [x] `exponential_decay_schedule`
- [x] `cosine_decay_schedule`
- [x] `constant_schedule`

## Layers

For now, just functional implementations resembling `torch.nn.functional` or `tf.nn`:

### Linear Layers

- [x] `dense`
- [x] `bilinear` - requires https://github.com/elixir-nx/exla/issues/182

### Convolutional Layers

- [x] `conv`
- [x] `conv_transpose`
- [x] `depthwise_conv`
- [x] `separable_conv2d`
- [x] `separable_conv3d`

### Pooling Layers

- [x] `avg_pool`
- [x] `max_pool`
- [x] `lp_pool`
- [x] `adaptive_avg_pool`
- [x] `adaptive_max_pool`
- [x] `adaptive_lp_pool`
- [x] `global_avg_pool`
- [x] `global_max_pool`
- [x] `global_lp_pool`

### Normalization Layers

- [x] `batch_norm`
- [x] `group_norm`
- [x] `instance_norm`
- [x] `layer_norm`

### Dropout Layers

- [x] `dropout`
- [x] `alpha_dropout`
- [x] `feature_alpha_dropout`
- [x] `spatial_dropout`

### Attention Layers

- [ ] `dot_product_attention` - requires https://github.com/elixir-nx/nx/issues/182
- [ ] `additive_attention` - requires `repeat`/`gather` on Nx

### Visual Layers

- [x] `resize`

We can drop off dimensional suffixes in favor of generic implementations too.

**Comment by seanmor5** [comment]
Closing in favor of smaller issues
